test/kt/single_pass_scan.cpp

```diff
@@ -0,0 +1,230 @@
+{
+    using NewKernelParam = make_kernel_param<KernelParam, usm_kt_kernel>;
+
+#if LOG_TEST_INFO
+    std::cout << "\t\ttest_usm<" << TypeInfo().name<T>() << ", " << USMAllocPresentation().name<_alloc_type>() << ">("
+              << size << ");" << std::endl;
+#endif
+    std::vector<T> expected(size);
+    generate_scan_data(expected.data(), size, 42);
+
+    TestUtils::usm_data_transfer<_alloc_type, T> dt_input(q, expected.begin(), expected.end());
+    TestUtils::usm_data_transfer<_alloc_type, T> dt_output(q, size);
+
+    std::inclusive_scan(expected.begin(), expected.end(), expected.begin(), bin_op);
+
+    oneapi::dpl::experimental::kt::gpu::inclusive_scan(q, dt_input.get_data(), dt_input.get_data() + size,
+                                                       dt_output.get_data(), bin_op, NewKernelParam{})
+        .wait();
+
+    std::vector<T> actual(size);
+    dt_output.retrieve_data(actual.begin());
+
+    std::string msg = "wrong results with USM, n: " + std::to_string(size);
+    EXPECT_EQ_N(expected.begin(), actual.begin(), size, msg.c_str());
+}
+
+template <typename T, typename BinOp, typename KernelParam>
+void
+test_sycl_iterators(sycl::queue q, std::size_t size, BinOp bin_op, KernelParam)
+{
+    constexpr oneapi::dpl::experimental::kt::kernel_param<KernelParam::data_per_workitem, KernelParam::workgroup_size>
+        param;
+#if LOG_TEST_INFO
+    std::cout << "\t\ttest_sycl_iterators<" << TypeInfo().name<T>() << ">(" << size << ");" << std::endl;
+#endif
+    std::vector<T> input(size);
+    std::vector<T> output(size);
+    generate_scan_data(input.data(), size, 42);
+    std::vector<T> ref(input);
+    std::inclusive_scan(std::begin(ref), std::end(ref), std::begin(ref), bin_op);
+    {
+        sycl::buffer<T> buf(input.data(), input.size());
+        sycl::buffer<T> buf_out(output.data(), output.size());
+        oneapi::dpl::experimental::kt::gpu::inclusive_scan(q, oneapi::dpl::begin(buf), oneapi::dpl::end(buf),
+                                                           oneapi::dpl::begin(buf_out), bin_op, param)
+            .wait();
+    }
+
+    std::string msg = "wrong results with oneapi::dpl::begin/end, n: " + std::to_string(size);
+    EXPECT_EQ_RANGES(ref, output, msg.c_str());
+}
+
+template <typename T, typename BinOp, typename KernelParam>
+void
+test_general_cases(sycl::queue q, std::size_t size, BinOp bin_op, KernelParam param)
+{
+    test_usm<T, sycl::usm::alloc::shared>(q, size, bin_op, param);
+    test_usm<T, sycl::usm::alloc::device>(q, size, bin_op, param);
+    test_sycl_iterators<T>(q, size, bin_op, param);
+#if _ENABLE_RANGES_TESTING
+    test_all_view<T>(q, size, bin_op, param);
+    test_buffer<T>(q, size, bin_op, param);
+#endif
+}
+
+template <typename T, typename KernelParam>
+void
+test_all_cases(sycl::queue q, std::size_t size, KernelParam param)
+{
+    test_general_cases<T>(q, size, std::plus<T>{}, param);
+    test_general_cases<T>(q, size, std::multiplies<T>{}, param);
```

Comment:
Could you make sure that there are no overflows? Perhaps, a different way of data generation or different sizes are needed.

Reply:
Here is what I suggest doing: 
[test_improvement.patch](https://github.com/oneapi-src/oneDPL/files/15027882/test_improvement.patch)

The patch improves data generation and checks if the kernel template can be executed.

Reply:
Thanks for the improvements. I incorporated them into this branch.